src/index.ts

```diff
@@ -0,0 +1,221 @@
+    let configParsed: ProjectOptions;
+    try {
+      const configContent = fs.readFileSync(filepath, 'utf-8');
+      configParsed = JSON.parse(configContent);
+    } catch (error) {
+      // swallow "not found" errors because they are so common / expected
+      if (error.code === 'ENOENT') {
+        logger.debug('no polymer config file found', {file: filepath});
+        return null;
+      }
+      // otherwise, throw an exception
+      throw error;
+    }
+
+    return configParsed;
+  }
+
+  /**
+   * Given an absolute file path to a polymer.json-like ProjectOptions object,
+   * return a new ProjectConfig instance created with those options.
+   */
+  static loadConfigFromFile(filepath: string): ProjectConfig {
+    let configParsed = ProjectConfig.loadOptionsFromFile(filepath);
+    if (!configParsed) {
+      return null;
+    }
+    return new ProjectConfig(configParsed);
+  }
+
+  /**
+   * constructor - given a ProjectOptions object, create the correct project
+   * configuration for those options. This involves setting the correct
+   * defaults, validating options, warning on deprecated options, and
+   * calculating some additional properties.
+   */
+  constructor(options: ProjectOptions) {
+    options = options || {};
+    options = fixDeprecatedOptions(options);
+
+    /**
+     * root
+     */
+    if (options.root) {
+      this.root = path.resolve(options.root);
+    } else {
+      this.root = process.cwd();
+    }
+
+    /**
+     * entrypoint
+     */
+    if (options.entrypoint) {
+      this.entrypoint = path.resolve(this.root, options.entrypoint);
+    } else {
+      this.entrypoint = path.resolve(this.root, 'index.html');
+    }
+
+    /**
+     * shell
+     */
+    if (options.shell) {
+      this.shell = path.resolve(this.root, options.shell);
+    }
+
+    /**
+     * fragments
+     */
+    if (options.fragments) {
+      this.fragments = options.fragments.map((e) => path.resolve(this.root, e));
+    }
+
+    /**
+     * dependencies
+     */
+    this.dependencies = (options.dependencies || [])
+        .map((glob) => resolveGlob(this.root, glob));
+
+    /**
+     * sources
+     */
+    this.sources = (options.sources || defaultSourceGlobs)
+        .map((glob) => resolveGlob(this.root, glob));
+    this.sources.push(this.entrypoint);
+    if (this.shell) {
+      this.sources.push(this.shell);
+    }
+    if (this.fragments) {
+      this.sources = this.sources.concat(this.fragments);
+    }
+
+    /**
+     * allFragments
+     */
+    this.allFragments = [];
+    // It's important that shell is first for document-ordering of imports
+    if (this.shell) {
+      this.allFragments.push(this.shell);
+    }
+    if (this.fragments) {
+      this.allFragments = this.allFragments.concat(this.fragments);
+    }
+    if (this.allFragments.length === 0) {
+      this.allFragments.push(this.entrypoint);
+    }
+  }
+
+  isFragment(filepath: string): boolean {
```

Comment:
easier said than done :/ I would prefer not to block unless we can come up with something on the spot that we can all agree on.

proposal: user definition of `fragments` remains the same, but the `fragments` property on `ProjectConfig` becomes `(options.fragments + options.shell) || (options.entrypoint)` (basically what allFragments already is). We can add a `userDefinedFragments` to reference the given fragments.

I admit its not perfect, but better than what we have today.